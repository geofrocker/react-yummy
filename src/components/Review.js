import React, { Component } from 'react';
import Since from 'react-since';
import  url, { http } from './config'

const Reviews = props =>
<div className="col-sm-12">
    <blockquote>
        <p>{props.content}.</p>
        <footer>By {props.created_by} about <cite title="Time"><Since date={ props.create_date }/></cite></footer>
    </blockquote>
</div>;
    
class Review extends Component {
    constructor(props){
        super(props)
        let recipe_id = this.props.match.params.recipe_id;
        this.state = {
            recipe_id:recipe_id,
            recipeData:[],
            message:'',
            color:'alert alert-danger',
            content:'',
            reviews:[],
            placeholder:false
        };
        
    }
    componentDidMount(){
        const {history} = this.props;
        if(!localStorage.getItem('isLoggedIn')){
            history.push('/login')
        }
        return http.get(`${url+this.state.recipe_id}`)
        .then((response) =>{
            if(response.data.Recipe_Item){
                this.setState({recipeData: response.data.Recipe_Item})
            }
        })
        .catch(xhr => {
            console.log(xhr)
        });  


    }

    componentWillMount(){
        return http.get(`${url}recipe/review/${this.state.recipe_id}`)
        .then((response) =>{
            console.log(response.data)
            this.setState({
                reviews: response.data.Review_list,
                placeholder:false
            })

        })
        .catch((xhr)=>{
            this.setState({
                placeholder: true
            })
        })
    }
    handleContentChange= (event) =>{
        this.setState({content: event.target.value})
    }

    upVote = (id) =>{
        return http.get(`${url}recipe/upvote/${this.state.recipe_id}`)
        .then((response)=>{
            let recipeData = this.state.recipeData;
            recipeData.upvotes = recipeData.upvotes +1
            this.setState({
                recipeData:recipeData,
                color:'alert alert-success',
                message:response.data.message,
                
            });
        })
        .catch(
            (xhr) => {
                this.setState({
                    color:'alert alert-danger',
                    message:xhr.response.data.message
                });
            }
        );
    };

    review = (e) =>{
        e.preventDefault();
        let postData = {
            content: this.state.content
        }
        return http.post(`${url}recipe/review/${this.state.recipe_id}`, postData)
        .then((response) => {
            console.log(response.data)
            let reviews = this.state.reviews;
            reviews.push(response.data.review)
            this.setState({
                placeholder:false,
                recipeData:response.data.Recipe,
                message:response.data.message,
                color:'alert alert-success',
                reviews:reviews,  
            })
           
            this.refs.content.value='';
        })
        .catch((xhr) => {
            this.setState({
                color:'alert alert-danger',
            })
        });
    }
    render(){
        const {placeholder, reviews, message, recipeData, color}=this.state

        let loadReviews;
        if (placeholder) {
            loadReviews = 
            <div>
                <p>This recipe has no reviews yet</p><br/>
            </div>
           
        }else {
            loadReviews =
            <div>
                {reviews.map(inf =>
                <Reviews key={inf.review_id} {...inf}/>
                )}
            </div>
        }
        return (
            <div className="Review">
                {this.state.message
                    ? <div className={color}>{message}</div>
                    : <div></div> 
                }
                <div className="jumbotron">
                    <h3>{recipeData.title}<a href="/" className="btn btn-primary pull-right">Go back</a></h3>
                    <em>Added by <span className="fa fa-user"></span> { recipeData.created_by } about <span className="fa fa-calendar"></span> <Since date={ recipeData.create_date } /></em>
                    <hr/>
                    <div>
                        <h3>Ingredients</h3>
                        <small>{recipeData.ingredients}</small>
                        <h3>Steps</h3>
                        <small>{recipeData.steps}</small><br/><br/>
                        <div className="btn-group">
                            <button type="button" onClick={() => this.upVote(recipeData.recipe_id)} className="btn btn-primary btn-xs">UpVote</button>
                            <button type="button" className="btn btn-default btn-xs"><span className="fa fa-thumbs-o-up"></span> Upvotes:{recipeData.upvotes}</button>
                            <button type="button" className="btn btn-default btn-xs"><span className="fa fa-comment-o"></span> Reviews:{recipeData.reviews}</button>
                        </div>
                    </div>
                </div>
                <div className="jumbotron review">
                    <p>Reviews</p>
                    {loadReviews}
                    <form onSubmit={this.review} id="review-form">
                            <div className="form-group">
                                <textarea className="form-control" placeholder="Enter your Review!" id="content" ref="content" onChange={this.handleContentChange} required/>
                            </div>
                            <div className="form-group">
                                <input type="submit" className="btn btn-primary pull-right" value="Add Review"></input>
                            </div>
                            <br/>
                    </form>
                </div>
            </div>

        );
        
    }
}

export default Review