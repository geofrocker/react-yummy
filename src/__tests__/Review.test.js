import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Review from '../components/review';

describe('add recipe component', () => {
    const props = {
        history: {
            push: () => {}
        },
        match:{
            params:{}
        }  
    }
    const wrapper = shallow(<Review {...props}/>);
    const preventDefault = jest.fn();
    
    it('renders the AddRecipe class', () => {
        expect(wrapper.find(".Review")).toHaveLength(1);
    });
    it('renders seven div jsx elements', () => {
        expect(wrapper.find("div")).toHaveLength(9);        
    });

    it('renders the the add review form and submit data', () =>{
        wrapper.setState({content:"awesome"});
        wrapper.find("#review-form").simulate('submit', {preventDefault});
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(preventDefault).toBeCalled();
    });

    it('render the inputs', () =>{
        expect(wrapper.find("#content")).toHaveLength(1);
    });

    it('renders two buttons jsx elements', () => {
        expect(wrapper.find(".btn")).toHaveLength(4);        
    });
    it('renders btn-group class jsx elements', () => {
        expect(wrapper.find(".btn-group")).toHaveLength(1);        
    });
    it('renders form-group class jsx elements', () => {
        expect(wrapper.find(".form-group")).toHaveLength(2);        
    });
    it('renders jumbotron class jsx elements', () => {
        expect(wrapper.find(".jumbotron")).toHaveLength(2);        
    });
   
})
