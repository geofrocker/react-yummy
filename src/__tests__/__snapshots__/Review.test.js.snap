// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add recipe component renders the the add review form and submit data 1`] = `
<div
  className="Review"
>
  <div />
  <div
    className="jumbotron"
  >
    <h3 />
    <em>
      Added by 
      <span
        className="fa fa-user"
      />
       
       about 
      <span
        className="fa fa-calendar"
      />
       
      <Since
        live={true}
      />
    </em>
    <hr />
    <div>
      <h3>
        Ingredients
      </h3>
      <small />
      <h3>
        Steps
      </h3>
      <small />
      <br />
      <br />
      <div
        className="btn-group"
      >
        <button
          className="btn btn-primary btn-xs"
          onClick={[Function]}
          type="button"
        >
          UpVote
        </button>
        <button
          className="btn btn-default btn-xs"
          type="button"
        >
          <span
            className="fa fa-thumbs-o-up"
          />
           Upvotes:
        </button>
        <button
          className="btn btn-default btn-xs"
          type="button"
        >
          <span
            className="fa fa-comment-o"
          />
           Reviews:
        </button>
      </div>
    </div>
  </div>
  <div
    className="jumbotron review"
  >
    <p>
      Reviews
    </p>
    <div />
    <form
      id="review-form"
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <textarea
          className="form-control"
          id="content"
          onChange={[Function]}
          placeholder="Enter your Review!"
          required={true}
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="btn btn-primary pull-right"
          type="submit"
          value="Add Review"
        />
      </div>
      <br />
    </form>
  </div>
</div>
`;
